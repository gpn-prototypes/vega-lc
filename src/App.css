html,
body,
#root {
  margin: 0;
  padding: 0;

  width: 100%;
  height: 100%;
}

.App {
  display: flex;
  flex-direction: column;
  align-items: flex-start;

  width: 100%;
  height: 100%;

  background: var(--color-bg-default);
}
