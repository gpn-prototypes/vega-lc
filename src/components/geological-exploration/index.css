.GeologicalExploration {
  width: 100%;
  height: calc(100vh - 32px);
}
