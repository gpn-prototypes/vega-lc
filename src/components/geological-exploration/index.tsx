import React, { useContext, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { IconProcessing, SplitPanes } from '@gpn-prototypes/vega-ui';

import { LogicConstructorWidget } from '../logic-constructor';
import { ObjectsGroupWidget } from '../objects-group';
import { ProjectStructureWidget } from '../project-structure';

import { cnGeologicalExploration } from './cn-geoexploration';

import './index.css';

import { ProjectContext } from '@/react-context/providers';
import { fetchActivitiesList } from '@/redux-store/activities/actions';
import { fetchGroupObjectList } from '@/redux-store/group-objects/actions';
import { fetchCanvasItemsData } from '@/redux-store/logic-constructor/actions';
import { setNotification } from '@/redux-store/notifications/actions';
import { fetchProjectStructureList } from '@/redux-store/project-structure/actions';

export const GeologicalExploration = (): React.ReactElement => {
  const dispatch = useDispatch();
  const { projectId, initialized } = useContext(ProjectContext);

  useEffect(() => {
    if (projectId) {
      dispatch(
        setNotification({
          message: 'Раздел находится в разработке',
          status: 'warning',
          icon: IconProcessing,
        }),
      );
    }
  }, [dispatch, projectId]);

  useEffect(() => {
    if (initialized) {
      dispatch(fetchActivitiesList());
      dispatch(fetchGroupObjectList());
      dispatch(fetchProjectStructureList());
      dispatch(fetchCanvasItemsData());
    }
  }, [dispatch, initialized]);

  return (
    <div className={cnGeologicalExploration()}>
      <SplitPanes split="vertical">
        <SplitPanes.Pane aria-label="trees" initialSize="260px" min="24px" max="260px">
          <SplitPanes split="horizontal">
            <SplitPanes.Pane aria-label="project-structure" min="50px">
              <ProjectStructureWidget />
            </SplitPanes.Pane>
            <SplitPanes.Pane aria-label="objects-groups" min="140px">
              <ObjectsGroupWidget />
            </SplitPanes.Pane>
          </SplitPanes>
        </SplitPanes.Pane>
        <SplitPanes.Pane aria-label="logic-constructor">
          <LogicConstructorWidget />
        </SplitPanes.Pane>
      </SplitPanes>
    </div>
  );
};
