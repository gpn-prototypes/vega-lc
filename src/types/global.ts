import ResizeObserver from 'resize-observer-polyfill';

declare global {
  interface Window {
    ResizeObserver: typeof ResizeObserver;
  }
}
