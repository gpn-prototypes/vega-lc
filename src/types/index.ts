export interface Identity {
  getToken(): Promise<string>;
}
