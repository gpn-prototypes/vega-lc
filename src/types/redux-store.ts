import React from 'react';
import { CanvasData, CanvasTree, TargetData, TreeItem } from '@gpn-prototypes/vega-ui';
import { CanvasView } from '@gpn-prototypes/vega-ui/dist/components/canvas/entities/CanvasView';

type NodeList = TreeItem[];

export declare type Content = {
  type: string;
  id: string;
  name: string;
};
export declare type Event = {
  id: string;
  name: string;
  content: Content[];
};
export declare type StepData = {
  id: string;
  name: string;
  events: Event[];
};

export type SplitPanesGridParams = {
  width: string;
  height?: string;
};

export type CanvasViewEntity = CanvasView;

export type StepContent = {
  activity?: {
    activityType: {
      vid: string;
    };
    name: string;
  };
  object?: {
    vid: string;
    name: string;
  };
};

export type Step = {
  name: string;
  vid: string;
  itemList: StepContent[];
};

export type CanvasElement = {
  id?: string;
  parentIds?: string[];
  childrenIds?: string[];
  data: CanvasData;
};

export type CanvasElements = {
  vid: string;
  children?: { vid: string }[];
  parents?: { vid: string }[];
  position: [number, number];
  nodeRef?: string | null;
  nodeType: 'root' | 'end' | 'domainObject';
  title?: string;
  width?: number;
};

export type ActivitiesState = {
  nodeList: NodeList;
  searchString: string | null;
  autoFocus: boolean;
  activitiesRef?: React.RefObject<HTMLElement> | null;
  draggingElements: TargetData[] | [];
  isDroppingOnExistingStep?: boolean;
  isActivitiesPanelOpen: boolean;
};

export type LogicConstructorState = {
  scenarioList?: Step[];
  canvasElements?: CanvasTree[];
  isStepEditorOpened?: boolean;
  canvasViewRef?: React.MutableRefObject<CanvasViewEntity | null>;
};

export type NewGroupParams = {
  isDynamic: boolean;
  name: string;
};

export type GroupObjectsState = {
  nodeList: NodeList;
  isDialogOpened: boolean;
  newGroupParams: NewGroupParams;
  draggingElements: TargetData[] | [];
};

export interface ProjectStructureQuery {
  query: string;
  tree: string[];
}

export type ProjectStructureState = {
  nodeList: NodeList;
  draggingElements: TargetData[] | [];
  projectStructureQuery?: ProjectStructureQuery;
};

export type VersionState = {
  version?: number;
};

export type StoreLC = {
  activities: ActivitiesState;
  groupObjects: GroupObjectsState;
  projectStructure: ProjectStructureState;
  logicConstructor: LogicConstructorState;
  version: VersionState;
  notifications: NotificationState;
};

export type NotificationStatus = 'alert' | 'normal' | 'system' | 'success' | 'warning' | undefined;

export type Notification = {
  message: string;
  status?: NotificationStatus;
  autoClose?: boolean;
  icon?: React.FC<any>;
};

export type NotificationState = {
  list: Notification[];
};
