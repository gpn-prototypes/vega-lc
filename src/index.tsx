import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { ApolloProvider } from '@apollo/client';
import { Root } from '@gpn-prototypes/vega-root';

import getStore from './redux-store/index';
import { App } from './App';
import client from './client';

import './App.css';

ReactDOM.render(
  <Root initialPortals={[{ name: 'modalRoot' }]} defaultTheme="dark">
    <ApolloProvider client={client}>
      <Provider store={getStore()}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    </ApolloProvider>
  </Root>,
  document.getElementById('root'),
);
