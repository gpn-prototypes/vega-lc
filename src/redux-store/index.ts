import { applyMiddleware, combineReducers, createStore, Store } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { persistStore } from 'redux-persist';
import thunkMiddleware from 'redux-thunk';

import activitiesReducer from './activities/reducer';
import groupObjectsReducer from './group-objects/reducer';
import logicConstructorReducer from './logic-constructor/reducer';
import notificationsReducer from './notifications/reducer';
import projectStructureReducer from './project-structure/reducer';
import versionReducer from './version/reducer';

import { StoreLC } from '@/types/redux-store';

const rootReducer = combineReducers({
  activities: activitiesReducer,
  groupObjects: groupObjectsReducer,
  projectStructure: projectStructureReducer,
  logicConstructor: logicConstructorReducer,
  version: versionReducer,
  notifications: notificationsReducer,
});

const getStore = (initialState?: StoreLC): Store =>
  createStore(
    rootReducer,
    initialState,
    composeWithDevTools(applyMiddleware(thunkMiddleware.withExtraArgument({}))),
  );

export const store = getStore();

export const persistor = persistStore(store);
