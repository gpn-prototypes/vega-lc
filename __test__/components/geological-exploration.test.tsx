import React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import ResizeObserver from 'resize-observer-polyfill';

import '../../src/types/global';

import { GeologicalExploration } from '../../src/components/geological-exploration';
import { ProjectContext } from '../../src/react-context/providers';
import { store } from '../../src/redux-store';

jest.mock('../../src/components/logic-constructor', () => {
  return {
    LogicConstructorWidget: () => {
      return <div />;
    },
  };
});

jest.mock('../../src/components/objects-group', () => {
  return {
    ObjectsGroupWidget: () => {
      return <div />;
    },
  };
});

jest.mock('../../src/components/project-structure', () => {
  return {
    ProjectStructureWidget: () => {
      return <div />;
    },
  };
});

const mockDispatch = jest.fn();

jest.mock('react-redux', () => {
  const originalModule = jest.requireActual('react-redux');

  return {
    ...originalModule,
    useDispatch: () => mockDispatch,
  };
});

beforeEach(() => {
  mockDispatch.mockClear();
});

const renderComponent = (projectId = '', initialized = false) => {
  return render(
    <ReduxProvider store={store}>
      <ProjectContext.Provider value={{ projectId, initialized }}>
        <GeologicalExploration />
      </ProjectContext.Provider>
    </ReduxProvider>,
  );
};

describe('Geological Exploration', () => {
  test('рендерится без ошибок', () => {
    window.ResizeObserver = ResizeObserver;

    const dom = renderer
      .create(
        <ReduxProvider store={store}>
          <GeologicalExploration />
        </ReduxProvider>,
      )
      .toJSON();
    expect(dom).toMatchSnapshot();
  });

  test('Основные компоненты присутствуют в DOM', () => {
    const { container } = render(
      <ReduxProvider store={store}>
        <GeologicalExploration />
      </ReduxProvider>,
    );

    const mainComponentLabels = ['project-structure', 'objects-groups', 'logic-constructor'];

    mainComponentLabels.forEach((label) =>
      expect(container.querySelector(`[aria-label="${label}"]`)).toBeInTheDocument(),
    );
  });

  test('вызывать загрузку данных проекта если он инициализирован', async () => {
    renderComponent('', true);

    expect(mockDispatch).toHaveBeenCalledTimes(4);
  });

  test('не вызывать загрузку данных проекта если он не инициализирован', () => {
    renderComponent('', false);

    expect(mockDispatch).not.toHaveBeenCalled();
  });
});
