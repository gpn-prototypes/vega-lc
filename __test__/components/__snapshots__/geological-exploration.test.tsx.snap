// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Geological Exploration рендерится без ошибок 1`] = `
<div
  className="GeologicalExploration"
>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
    role="tree"
  >
    <div
      aria-label="trees"
      className="VegaSplitPanes__Pane"
      role="treeitem"
      style={
        Object {
          "flexGrow": 0,
          "maxWidth": "260px",
          "minWidth": "24px",
          "width": "260px",
        }
      }
    >
      <div
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        role="tree"
      >
        <div
          aria-label="project-structure"
          className="VegaSplitPanes__Pane"
          role="treeitem"
          style={
            Object {
              "flex": "1",
              "maxHeight": "calc(100% - 2px * 1)",
              "minHeight": "50px",
            }
          }
        >
          <div />
        </div>
        <div
          aria-current={false}
          aria-label="Изменить размер"
          className="VegaSplitPanes__Resizer VegaSplitPanes__Resizer_split_horizontal"
          onMouseDown={[Function]}
          role="presentation"
          tabIndex={-1}
        />
        <div
          aria-label="objects-groups"
          className="VegaSplitPanes__Pane"
          role="treeitem"
          style={
            Object {
              "flex": "1",
              "maxHeight": "calc(100% - 2px * 1)",
              "minHeight": "140px",
            }
          }
        >
          <div />
        </div>
      </div>
    </div>
    <div
      aria-current={false}
      aria-label="Изменить размер"
      className="VegaSplitPanes__Resizer VegaSplitPanes__Resizer_split_vertical"
      onMouseDown={[Function]}
      role="presentation"
      tabIndex={-1}
    />
    <div
      aria-label="logic-constructor"
      className="VegaSplitPanes__Pane"
      role="treeitem"
      style={
        Object {
          "flex": "1",
          "maxWidth": "calc(100% - 2px * 1)",
          "minWidth": "0",
        }
      }
    >
      <div />
    </div>
  </div>
</div>
`;
